### PR TITLE
feat(langgraph): Adds stores & managed values (memory)

### PR DESCRIPTION
**Pull request by bracesproul**
TODO: add tests that verifies managed values get passed through, but aren't stored in checkpointer

### CODE REVIEW

File: libs/langgraph/src/managed/shared_value.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    await instance.tick();
+    return instance as unknown as ManagedValue<Value>;
+  }
+
+  static on(scope: string): ConfiguredManagedValue<Value> {
+    return {
+      cls: SharedValue,
+      params: {
+        scope,
+        key: {},
```
Comment (bracesproul) [inline libs/langgraph/src/managed/shared_value.ts:70]: in py this is typed as a string, but defaults to an object. figure out which it should be.

File: libs/langgraph/src/channels/base.ts [hunk 2]
```diff
@@ -6,6 +6,10 @@ import {
 } from "@langchain/langgraph-checkpoint";
 import { EmptyChannelError } from "../errors.js";
 
+export function isBaseChannel(obj: unknown): obj is BaseChannel {
+  return obj != null && typeof (obj as BaseChannel).lc_graph_name === "string";
```
Comment (jacoblee93) [inline libs/langgraph/src/channels/base.ts:10]: I don't think we should do this as a rule (`Send` has this populated) - can we add some other field?

File: libs/langgraph/src/channels/base.ts [hunk 3]
```diff
@@ -75,11 +79,15 @@ export function emptyChannels<Cc extends Record<string, BaseChannel>>(
   channels: Cc,
   checkpoint: ReadonlyCheckpoint
 ): Cc {
+  const filteredChannels = Object.fromEntries(
```
Comment (jacoblee93) [inline libs/langgraph/src/channels/base.ts:82]: This removes managed values?

File: examples/how-tos/shared-state.ipynb [hunk 4]
```diff
@@ -0,0 +1,779 @@
...
+    "    throw new Error(\"Expected an AI message with tool calls.\");\n",
+    "  }\n",
+    "\n",
+    "  return (lastMessage as AIMessage).tool_calls?.length ? \"update_memory\" : END;\n",
+    "};\n",
+    "\n",
+    "const updateMemory = (\n",
+    "  state: typeof AgentAnnotation.State\n",
+    "): Partial<typeof AgentAnnotation.State> => {\n",
+    "  const toolCalls: ToolMessage[] = [];\n",
```
Comment (jacoblee93) [inline examples/how-tos/shared-state.ipynb:369]: These aren't tool calls, should be called `toolResponseMessages` or something

File: examples/how-tos/shared-state.ipynb [hunk 5]
```diff
@@ -0,0 +1,779 @@
...
+   "cell_type": "code",
+   "execution_count": 2,
+   "id": "337ee88f",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Result from running graph with input \"hi\":\n",
```
Comment (jacoblee93) [inline examples/how-tos/shared-state.ipynb:71]: This below example could use a few more comments

File: examples/how-tos/shared-state.ipynb [hunk 6]
```diff
@@ -0,0 +1,779 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b8c416fa-086a-491d-a7d3-57091f6413e3",
+   "metadata": {},
+   "source": [
+    "## Run graph on a different thread\n",
+    "\n",
+    "We can now run the graph on a different thread and see that it remembers facts about the user (specifically that the user likes pepperoni pizza and lives in SF):"
```
Comment (jacoblee93) [inline examples/how-tos/shared-state.ipynb:635]: Isn't this what we just did above?

File: libs/langgraph/src/managed/base.ts [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export abstract class ManagedValue<Value = any> {
+  runtime: boolean = false;
+
+  config: RunnableConfig;
+
+  private _promises: Promise<unknown>[] = [];
+
+  ls_is_managed_value = true;
```
Comment (jacoblee93) [inline libs/langgraph/src/managed/base.ts:15]: Call this `lg_is_managed_value` - `ls` is LangSmith

File: libs/langgraph/src/managed/base.ts [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+  constructor(config: RunnableConfig, _params?: ManagedValueParams) {
+    this.config = config;
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  static async initialize<Value = any>(
```
Comment (jacoblee93) [inline libs/langgraph/src/managed/base.ts:22]: Just make this an abstract method?

File: libs/langgraph/src/managed/shared_value.ts [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+      this.ns = null;
+    } else if (config.configurable?.[this.scope]) {
+      const scopeValue = config.configurable[this.scope];
+      let scopedValueString = "";
+      try {
+        scopedValueString =
+          typeof scopeValue === "string"
+            ? scopeValue
+            : JSON.stringify(scopeValue);
+      } catch (_) {
```
Comment (jacoblee93) [inline libs/langgraph/src/managed/shared_value.ts:45]: When would this fail?

File: libs/langgraph/src/managed/base.ts [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+    _config: RunnableConfig,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    _args?: any
+  ): Promise<ManagedValue<Value>> {
+    throw new Error("Not implemented");
+  }
+
+  abstract call(step: number): Value;
```
Comment (jacoblee93) [inline libs/langgraph/src/managed/base.ts:30]: Why not make this a runnable?

File: libs/langgraph/src/pregel/algo.ts [hunk 11]
```diff
@@ -143,7 +189,10 @@ export function _applyWrites<Cc extends Record<string, BaseChannel>>(
   tasks: WritesProtocol<keyof Cc>[],
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   getNextVersion?: (version: any, channel: BaseChannel) => any
-): void {
+): Record<string, PendingWriteValue[]> {
+  const filteredChannels = Object.fromEntries(
```
Comment (jacoblee93) [inline libs/langgraph/src/pregel/algo.ts:193]: See above, I *think* this is filtering out managed values but more clarity around naming would be nice.

File: libs/langgraph/src/pregel/algo.ts [hunk 12]
```diff
@@ -356,22 +419,33 @@ export function _prepareNextTasks<
...
+                    step,
+                    (items: [keyof Cc, unknown][]) => writes.push(...items),
+                    processes,
+                    channels,
+                    managed,
+                    writes_
+                  ),
+                [CONFIG_KEY_READ]: (
+                  select_: Array<keyof Cc> | keyof Cc,
+                  fresh_: boolean = false
```
Comment (jacoblee93) [inline libs/langgraph/src/pregel/algo.ts:434]: Why are these params suffixed?

Also what do they represent?

File: libs/langgraph/src/pregel/index.ts [hunk 13]
```diff
@@ -775,7 +875,12 @@ export class Pregel<
       await runManager?.handleChainError(e);
       throw e;
     } finally {
-      await Promise.all(loop?.checkpointerPromises ?? []);
+      this.store?.stop();
+      loop?.store?.stop();
```
Comment (jacoblee93) [inline libs/langgraph/src/pregel/index.ts:879]: Why would there be two different values here?

File: libs/langgraph/src/store/batch.ts [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+
+  /**
+   * Runs the task that processes queued operations in batches.
+   * This method runs continuously until the store is stopped,
+   * or the process is terminated.
+   * @returns {Promise<void>} A promise that resolves when the task is complete.
+   */
+  private async runTask(): Promise<void> {
+    while (this.running) {
+      await new Promise((resolve) => {
```
Comment (jacoblee93) [inline libs/langgraph/src/store/batch.ts:125]: What is this for?

Comment (jacoblee93) [reply libs/langgraph/src/store/batch.ts:125]: If it is necessary, add a comment explaining why. But it seems like a race condition waiting to happen.

File: libs/langgraph/src/store/batch.ts [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+    this.running = false;
+  }
+
+  /**
+   * Runs the task that processes queued operations in batches.
+   * This method runs continuously until the store is stopped,
+   * or the process is terminated.
+   * @returns {Promise<void>} A promise that resolves when the task is complete.
+   */
+  private async runTask(): Promise<void> {
```
Comment (jacoblee93) [inline libs/langgraph/src/store/batch.ts:123]: This is a bad name, you're basically calling it `doThing()`

File: libs/langgraph/src/store/batch.ts [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+    while (this.running) {
```
Comment (jacoblee93) [inline libs/langgraph/src/store/batch.ts:124]: Is there really no better way to do this? For example, can the loop call this at the end of each tick?

**Comment by jacoblee93** [comment]
All tests have been ported from Python?

**Review by jacoblee93 — COMMENTED** [review]
x

File: libs/langgraph/src/pregel/loop.ts [hunk 17]
```diff
@@ -232,6 +257,14 @@ export class PregelLoop {
     this.checkpointerPromises.push(this._checkpointerChainedPromise);
   }
 
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  protected async updateMv(key: string, values: any[]) {
```
Comment (jacoblee93) [inline libs/langgraph/src/pregel/loop.ts:261]: nit: avoid abbreviations like this - just `updateManagedValues`

File: libs/langgraph/src/managed/base.ts (hunk 8 above)
Comment (bracesproul) [reply libs/langgraph/src/managed/base.ts:22]: Can't make static methods abstract

File: libs/langgraph/src/managed/shared_value.ts (hunk 9 above)
Comment (bracesproul) [reply libs/langgraph/src/managed/shared_value.ts:45]: Good catch, I don't think it will. removed

File: libs/langgraph/src/pregel/algo.ts (hunk 11 above)
Comment (bracesproul) [reply libs/langgraph/src/pregel/algo.ts:193]: it is, will rename

File: libs/langgraph/src/pregel/algo.ts (hunk 12 above)
Comment (bracesproul) [reply libs/langgraph/src/pregel/algo.ts:434]: suffix to show we shouldn't do anything with them besides passing args through. They're the two args which we aren't populating, but this func requires so we create a `partial` func. No change from what we had before, except a new type error occurred with the existing implementation. (plus I think naming conflicts)

File: libs/langgraph/src/store/batch.ts (hunk 16 above)
Comment (bracesproul) [reply libs/langgraph/src/store/batch.ts:124]: ya thats a good idea, pushed fix

File: libs/langgraph/src/store/batch.ts (hunk 15 above)
Comment (bracesproul) [reply libs/langgraph/src/store/batch.ts]: renamed to `processBatchQueue`

File: libs/langgraph/src/pregel/index.ts (hunk 13 above)
Comment (bracesproul) [reply libs/langgraph/src/pregel/index.ts]: fixed

File: libs/langgraph/src/managed/base.ts [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+            throw error;
+          }
+        }
+      }
+    }
+  }
+}
+
+export function isManagedValue(value: unknown): value is typeof ManagedValue {
+  if (typeof value === "object" && value && "lg_is_managed_value" in value) {
```
Comment (jacoblee93) [inline libs/langgraph/src/managed/base.ts:169]: Don't forget to update name here

**Pull request merged**